File: src/spdx.test.ts
import { test } from "node:test";

import { expect } from "expect";

import { isValidSpdxExpression } from "./spdx.ts";

test("single SPDX ids are accepted", () => {
  for (const id of ["MIT", "Apache-2.0", "GPL-3.0-or-later", "EPL-2.0", "BSD-3-Clause", "ISC"]) {
    expect(isValidSpdxExpression(id)).toBe(true);
  }
});

test("compound expressions, +, parentheses and WITH are accepted", () => {
  expect(isValidSpdxExpression("MIT OR Apache-2.0")).toBe(true);
  expect(isValidSpdxExpression("(MIT OR Apache-2.0)")).toBe(true);
  expect(isValidSpdxExpression("Apache-2.0 AND MIT")).toBe(true);
  expect(isValidSpdxExpression("GPL-2.0-or-later+")).toBe(true);
  expect(isValidSpdxExpression("GPL-2.0-only WITH Classpath-exception-2.0")).toBe(true);
  expect(isValidSpdxExpression("(MIT OR (Apache-2.0 AND ISC))")).toBe(true);
});

test("free text, unknown ids and malformed expressions are rejected", () => {
  expect(isValidSpdxExpression("The Apache Software License, Version 2.0")).toBe(false);
  expect(isValidSpdxExpression("Definitely-Not-A-License")).toBe(false);
  expect(isValidSpdxExpression("")).toBe(false);
  expect(isValidSpdxExpression("MIT OR")).toBe(false);
  expect(isValidSpdxExpression("OR MIT")).toBe(false);
  expect(isValidSpdxExpression("(MIT")).toBe(false);
  expect(isValidSpdxExpression("MIT Apache-2.0")).toBe(false); // missing operator
  expect(isValidSpdxExpression("MIT WITH Bogus-exception")).toBe(false);
});

test("WITH clauses, nesting and unbalanced parentheses", () => {
  // an operand continues correctly after a WITH clause
  expect(isValidSpdxExpression("MIT WITH Classpath-exception-2.0 OR Apache-2.0")).toBe(true);
  expect(isValidSpdxExpression("(MIT OR (Apache-2.0 AND ISC))")).toBe(true);
  expect(isValidSpdxExpression("GPL-2.0-or-later+")).toBe(true);

  expect(isValidSpdxExpression("MIT WITH")).toBe(false); // WITH with no following exception
  expect(isValidSpdxExpression("MIT WITH MIT")).toBe(false); // exception slot is a license id
  expect(isValidSpdxExpression("AND MIT")).toBe(false); // operator where an operand is expected
  expect(isValidSpdxExpression("MIT AND OR Apache-2.0")).toBe(false); // two operators in a row
  expect(isValidSpdxExpression("(MIT))")).toBe(false); // extra close paren
  expect(isValidSpdxExpression("((MIT)")).toBe(false); // unclosed group
});

File: togo/internal/config/spdx_test.go
package config

import "testing"

func TestIsValidSpdxExpression(t *testing.T) {
	valid := []string{
		"MIT",
		"Apache-2.0",
		"(MIT OR Apache-2.0)",
		"GPL-2.0-only WITH Classpath-exception-2.0",
		"MIT AND (Apache-2.0 OR BSD-3-Clause)",
		"LGPL-2.1+",
	}
	for _, e := range valid {
		if !IsValidSpdxExpression(e) {
			t.Errorf("expected %q to be valid", e)
		}
	}
	invalid := []string{
		"",
		"The Apache Software License, Version 2.0",
		"MIT OR",
		"(MIT",
		"MIT)",
		"NotARealLicense",
		"MIT WITH NotAnException",
	}
	for _, e := range invalid {
		if IsValidSpdxExpression(e) {
			t.Errorf("expected %q to be invalid", e)
		}
	}
}

func TestSpdxBranchCoverage(t *testing.T) {
	valid := []string{
		"MIT WITH Classpath-exception-2.0 OR Apache-2.0", // an operand continues after a WITH clause
		"(MIT OR (Apache-2.0 AND ISC))",                  // nested groups
		"GPL-2.0-or-later+",                              // or-later suffix
	}
	for _, e := range valid {
		if !IsValidSpdxExpression(e) {
			t.Errorf("expected %q to be valid", e)
		}
	}
	invalid := []string{
		"MIT WITH",              // WITH with no following exception
		"MIT WITH MIT",          // the exception slot holds a license id, not an exception
		"AND MIT",               // an operator where an operand is expected
		"MIT AND OR Apache-2.0", // two operators in a row
		"(MIT))",                // unbalanced: an extra close paren
		"((MIT)",                // unbalanced: an unclosed group
	}
	for _, e := range invalid {
		if IsValidSpdxExpression(e) {
			t.Errorf("expected %q to be invalid", e)
		}
	}
}
